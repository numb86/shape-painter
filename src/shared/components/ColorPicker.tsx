import React, {useState, useEffect, useCallback, useRef} from 'react';
import {SketchPicker, ColorResult} from 'react-color';
import styled from 'styled-components';

type Props = {
  color: string;
  setColor: (selectedColor: string) => void;
  disable: boolean;
  closePickerCallback?: () => void;
};

type PopoverProps = Pick<
  Required<Props>,
  'color' | 'setColor' | 'closePickerCallback'
> & {
  setVisibility: React.Dispatch<React.SetStateAction<boolean>>;
};

const Popover = ({
  color,
  setColor,
  closePickerCallback,
  setVisibility,
}: PopoverProps) => {
  const selectColor = useCallback(
    (result: ColorResult) => {
      setColor(result.hex);
    },
    [setColor]
  );

  const willUnMount = useRef(false);

  const closePicker = useCallback(() => {
    willUnMount.current = true;
    setVisibility(false);
  }, [setVisibility]);

  useEffect(() => {
    return () => {
      if (willUnMount.current) {
        closePickerCallback();
      }
    };
  }, [closePickerCallback]);

  return (
    <PopoverStyled>
      <Cover onClick={closePicker} data-testid="cover" />
      <SketchPicker color={color} onChangeComplete={selectColor} />
    </PopoverStyled>
  );
};

const ColorPicker = ({
  color,
  setColor,
  disable,
  closePickerCallback = () => {},
}: Props) => {
  const [visibility, setVisibility] = useState(false);

  const openPicker = useCallback(() => {
    if (disable) return;
    setVisibility(true);
  }, [disable]);

  return (
    <>
      <Frame type="button" disable={disable} onClick={openPicker}>
        <Swatch color={color} disable={disable} />
      </Frame>
      {visibility && (
        <Popover
          color={color}
          setColor={setColor}
          setVisibility={setVisibility}
          closePickerCallback={closePickerCallback}
        />
      )}
    </>
  );
};

export default ColorPicker;

const Swatch = styled.div<{color: string; disable: boolean}>`
  width: 36px;
  height: 14px;
  border-radius: 2px;
  background-color: ${(props) => props.color};
  visibility: ${(props) => (props.disable ? 'hidden' : 'visible')};
`;

const Frame = styled.button<{disable: boolean}>`
  border: none;
  padding: 5px;
  border-radius: 1px;
  box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.1);
  display: inline-block;
  cursor: ${(props) => (props.disable ? 'not-allowed' : 'pointer')};
  background-color: #fff;
  background-image: ${(props) =>
    props.disable
      ? `linear-gradient(
    to right top,
    rgba(255, 255, 255, 0) 46%,
    #f00 49%,
    #f00 51%,
    rgba(255, 255, 255, 0) 54%
  )`
      : 'none'};
`;

const PopoverStyled = styled.div`
  position: absolute;
`;

const Cover = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: #000;
  opacity: 0.2;
`;
