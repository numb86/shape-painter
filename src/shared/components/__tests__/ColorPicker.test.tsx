// This test require `canvas`
// https://github.com/Automattic/node-canvas

import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import 'jest-styled-components';
import assert from 'power-assert';

import {ColorPicker} from '../ColorPicker';

describe('ColorPicker', () => {
  it('Swatch background color is props.color', () => {
    const {container: c1} = render(
      <ColorPicker color="#000" setColor={() => {}} disable={false} />
    );
    expect(c1.firstChild).toMatchSnapshot();

    const {container: c2} = render(
      <ColorPicker color="#f00" setColor={() => {}} disable={false} />
    );
    expect(c2.firstChild).toMatchSnapshot();

    const {container: c3} = render(
      <ColorPicker color="#aaa" setColor={() => {}} disable={false} />
    );
    expect(c3.firstChild).toMatchSnapshot();
  });

  describe('Click Frame', () => {
    it('SketchPicker opens when props.disable is false', () => {
      render(<ColorPicker color="#000" setColor={() => {}} disable={false} />);
      fireEvent.click(screen.getByRole('button'));
      assert(screen.queryByTestId('cover'));
    });

    it('SketchPicker does not open when props.disable is true', () => {
      render(<ColorPicker color="#000" setColor={() => {}} disable />);
      fireEvent.click(screen.getByRole('button'));
      assert(screen.queryByTestId('cover') === null);
    });
  });

  describe('Click Cover', () => {
    it('Close SketchPicker', () => {
      render(<ColorPicker color="#000" setColor={() => {}} disable={false} />);
      fireEvent.click(screen.getByRole('button'));
      assert(screen.queryByTestId('cover'));
      fireEvent.click(screen.getByTestId('cover'));
      assert(screen.queryByTestId('cover') === null);
    });

    it('Execute props.closePickerCallback', () => {
      const closePickerCallback = jest.fn();
      render(
        <ColorPicker
          color="#000"
          setColor={() => {}}
          disable={false}
          closePickerCallback={closePickerCallback}
        />
      );

      fireEvent.click(screen.getByRole('button'));
      expect(closePickerCallback).toHaveBeenCalledTimes(0);

      fireEvent.click(screen.getByTestId('cover'));
      expect(closePickerCallback).toHaveBeenCalledTimes(1);
    });
  });
});
