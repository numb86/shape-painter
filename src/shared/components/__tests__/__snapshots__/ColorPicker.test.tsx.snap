// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker Swatch background color is props.color 1`] = `
.c1 {
  width: 36px;
  height: 14px;
  border-radius: 2px;
  background-color: #000;
  visibility: visible;
}

.c0 {
  border: none;
  padding: 5px;
  border-radius: 1px;
  box-shadow: 0 0 0 1px rgba(0,0,0,0.1);
  display: inline-block;
  cursor: pointer;
  background-color: #fff;
  background-image: none;
}

<button
  class="c0"
  type="button"
>
  <div
    class="c1"
    color="#000"
  />
</button>
`;

exports[`ColorPicker Swatch background color is props.color 2`] = `
.c1 {
  width: 36px;
  height: 14px;
  border-radius: 2px;
  background-color: #f00;
  visibility: visible;
}

.c0 {
  border: none;
  padding: 5px;
  border-radius: 1px;
  box-shadow: 0 0 0 1px rgba(0,0,0,0.1);
  display: inline-block;
  cursor: pointer;
  background-color: #fff;
  background-image: none;
}

<button
  class="c0"
  type="button"
>
  <div
    class="c1"
    color="#f00"
  />
</button>
`;

exports[`ColorPicker Swatch background color is props.color 3`] = `
.c1 {
  width: 36px;
  height: 14px;
  border-radius: 2px;
  background-color: #aaa;
  visibility: visible;
}

.c0 {
  border: none;
  padding: 5px;
  border-radius: 1px;
  box-shadow: 0 0 0 1px rgba(0,0,0,0.1);
  display: inline-block;
  cursor: pointer;
  background-color: #fff;
  background-image: none;
}

<button
  class="c0"
  type="button"
>
  <div
    class="c1"
    color="#aaa"
  />
</button>
`;
